### DISCUSSION TITLE
2 questions about CallTraceStorage

### DISCUSSION THREAD

**Discussion by yanglong1010** [async-profiler/async-profiler#1087] [General]
hi,

I have two questions that I'm confused about, and I hope to get answers.

1. Why `CallTraceStorage::findCallTrace` does not search stacktrace recursively through `LongHashTable._prev` ?

https://github.com/async-profiler/async-profiler/blob/8da7831bb1fbd5943ab1f5c5b54f35a18f2a7bf5/src/callTraceStorage.cpp#L251

If the `CallTraceStorage._current_table` has previous table, and the previous table has previous table (i.e. _current_table -> previous A -> previous B), only search `previous A` can lead to stack traces in `previous B` can not be found, then will storeCallTrace again. The result is that a stack trace is stored twice (one in `_current_table` and one in `previous B`).

2. If two stack traces have the same hash (calcHash), I don't see any handling of hash collisions, why ?

Thanks.

**Comment by apangin** [comment]
> The result is that a stack trace is stored twice

Yes, this is an intended behavior. Looking into a previous table is just an optimization to save some space in a common case. Traversing unlimited number of tables will degrade hash table performance for no good reason.

> I don't see any handling of hash collisions, why ?

See #931

**Reply by yanglong1010** [reply]
Thanks a lot.
